src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/CSharpConvertToRecordRefactoringProvider.cs

```diff
@@ -0,0 +1,791 @@
+                IsDotEqualsInvocation(returnOperation.ChildOperations.First(), type);
+        }
+
+        // must be of form !(c1 == c2) or !c1.Equals(c2)
+        private static bool IsDefaultNotEqualsOperator(
+            OperatorDeclarationSyntax operatorDeclaration,
+            INamedTypeSymbol type,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            if (semanticModel.GetOperation(operatorDeclaration, cancellationToken) is not IMethodBodyOperation operation)
+            {
+                return false;
+            }
+
+            var body = operation.BlockBody ?? operation.ExpressionBody;
+            // if any of these conditions are false we short-circuit false
+            if (!(body != null &&
+                body.Operations.IsSingle() &&
+                operation.BlockBody?.Operations[0] is IReturnOperation returnOperation &&
+                returnOperation.ChildOperations.Count == 1 &&
+                returnOperation.ChildOperations.First() is IUnaryOperation notOp &&
+                notOp.Syntax.IsKind(SyntaxKind.LogicalNotExpression) &&
+                notOp.ChildOperations.Count == 1))
+            {
+                return false;
+            }
+
+            if (IsDotEqualsInvocation(notOp.ChildOperations.First(), type))
+            {
+                return true;
+            }
+
+            if (!(notOp.ChildOperations.First() is IBinaryOperation equalsOp &&
+                equalsOp.Syntax.IsKind(SyntaxKind.EqualsExpression)))
+            {
+                return false;
+            }
+
+            // either could potentially be a param reference or an implicitly cast param reference
+            // based on the definition of the == operator
+            var left = GetParamFromArgument(equalsOp.LeftOperand);
+            var right = GetParamFromArgument(equalsOp.RightOperand);
+            return (left != null && right != null && !left.Equals(right));
+        }
+
+        // matches form
+        // c1.Equals(c2)
+        // where one of c1 or c2 is the given type
+        // and the other is the same type or object
+        private static bool IsDotEqualsInvocation(IOperation operation, INamedTypeSymbol type)
+        {
+            // any of these being false will short-circuit false
+            if (!(operation is IInvocationOperation invocation &&
+                invocation.TargetMethod.Name == nameof(Equals) &&
+                invocation.ChildOperations.Count == 2 &&
+                invocation.ChildOperations.First() is IParameterReferenceOperation { Parameter: var invokedOn } &&
+                invocation.Arguments.Length == 1))
+            {
+                return false;
+            }
+
+            var param = GetParamFromArgument(invocation.Arguments[0].Value);
+
+            return param != null && !invokedOn.Equals(param);
+        }
+
+        private static IParameterSymbol? GetParamFromArgument(IOperation arg)
```

Comment:
i think you can replace this with `WalkDownConversion`, then see if yo uhave an IParameterReferenceOp